reverse_iterator.hpp

```diff
@@ -0,0 +1,224 @@
+        template<class OtherIter, class>
+        reverse_iterator<Iter> reverse_iterator<Iter, unused>::operator=(
+                const reverse_iterator<OtherIter> &other) {
+            _iter = other._iter;
+            return *this;
+        }
+
+
+        template<class Iter, class unused>
+        typename reverse_iterator<Iter, unused>::reference
+        reverse_iterator<Iter, unused>::operator*() const {
+            return *_iter;
+        }
+
+        template<class Iter, class unused>
+        typename reverse_iterator<Iter, unused>::pointer
+        reverse_iterator<Iter, unused>::operator->() const {
+            return std::addressof(operator*());
+        }
+
+        template<class Iter, class unused>
+        reverse_iterator<Iter, unused> &reverse_iterator<Iter, unused>::operator++() {
+            --_iter;
+            return *this;
+        }
+
+        template<class Iter, class unused>
+        reverse_iterator<Iter, unused> reverse_iterator<Iter, unused>::operator++(int) {
+            reverse_iterator copy = *this;
+            ++(*this);
+            return copy;
+        }
+
+        template<class Iter, class unused>
+        reverse_iterator<Iter, unused> &reverse_iterator<Iter, unused>::operator--() {
+            ++_iter;
+            return *this;
+        }
+
+        template<class Iter, class unused>
+        reverse_iterator<Iter, unused> reverse_iterator<Iter, unused>::operator--(int) {
+            reverse_iterator copy = *this;
+            --(*this);
+            return copy;
+        }
+
+        template<class Iter, class unused>
+        typename reverse_iterator<Iter, unused>::iterator_type
+        reverse_iterator<Iter, unused>::base() const {
+            iterator_type base = _iter;
+            ++base;
+            return base;
+        }
+
+        template<class Iter, class unused>
+        bool reverse_iterator<Iter, unused>::operator==(const reverse_iterator &other) {
+            return (_iter == other._iter);
+        }
+
+        template<class Iter, class unused>
+        reverse_iterator<Iter, unused>
+        reverse_iterator<Iter, unused>::operator[](difference_type index) {
+            if constexpr (lghazarosyan::is_base_of_v<std::random_access_iterator_tag, typename Iter::iterator_category>){
+                return *this + index;
+            }
+            else{
+                static_assert(lghazarosyan::is_base_of_v<std::random_access_iterator_tag, typename Iter::iterator_category>,"Iterator must be random_access for operator []");
+            }
+        }
+
+        template<class Iter, class unused>
+        reverse_iterator<Iter, unused> &
+        reverse_iterator<Iter, unused>::operator-=(difference_type index) {
+            if constexpr (lghazarosyan::is_base_of_v<std::random_access_iterator_tag, typename Iter::iterator_category>){
+                reverse_iterator<Iter>::_iter += index;
+                return *this;
+            }
+            else{
```

Comment:
remove else, if it enters `if` then it should return

Reply:
No this is not ordinary if . This is constexpr , and I don't want to static_assert in every instantiation (removing else will cause it).

Reply:
ok